File: reverse_nodes_in_k_group.cpp
//
// Created by Harsha Vardhan on 14/01/22.
//

#include <iostream>
#include <stack>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;

    ListNode() : val(0), next(nullptr) {}

    ListNode(int x) : val(x), next(nullptr) {}

    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode* reverseK(ListNode* head, int k, int j) {
    if (j == 0) return head;
    ListNode* next = NULL;
    ListNode* curr = head;
    ListNode* prev = NULL;
    int i = 0;
    while (i < k) {
        next = curr->next;
        curr->next = prev;
        prev = curr;
        curr = next;
        i++;
    }
    j--;
    head->next = reverseK(curr, k, j);
    return prev;
}
int main() {
    ListNode *z = NULL;
    ListNode *prev = NULL;
    int n = 9;
    for (int i = 0; i < n; i++) {
        z = new ListNode(n - i, prev);
        prev = z;
    }
    int k = 3;
    reverseK(z, k, n/k);
    return 0;
}

File: go/linked-lists/reverse_nodes_in_k_group.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseHelper(head *ListNode, k int) (*ListNode, *ListNode) {
	curr := head
	next := head
	var prev *ListNode

	for k > 0 {
		next = curr.Next
		curr.Next = prev
		prev = curr
		curr = next
		k--
	}

	return prev, curr
}

func reverseKGroupHelper(head *ListNode, k int, numGroups int) *ListNode {
	if numGroups == 0 {
		return head
	}
	prev, curr := reverseHelper(head, k)
	head.Next = reverseKGroupHelper(curr, k, numGroups-1)
	return prev
}

func reverseKGroup(head *ListNode, k int) *ListNode {
	tmp := head
	n := 0
	for tmp != nil {
		n++
		tmp = tmp.Next
	}
	return reverseKGroupHelper(head, k, n/k)
}
